libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt

```diff
@@ -0,0 +1,139 @@
+import com.mapbox.navigation.core.infra.factories.createNavigationStatus
+import com.mapbox.navigation.core.infra.factories.createVoiceInstruction
+import com.mapbox.navigation.core.infra.recorders.RouteProgressObserverRecorder
+import com.mapbox.navigation.core.infra.recorders.VoiceInstructionsObserverRecorder
+import com.mapbox.navigation.core.trip.service.TripService
+import com.mapbox.navigation.utils.internal.JobControl
+import com.mapbox.navigation.utils.internal.ThreadController
+import com.mapbox.navigator.NavigationStatusOrigin
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.spyk
+import junit.framework.Assert.assertEquals
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.InternalCoroutinesApi
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.annotation.Config
+
+@InternalCoroutinesApi
+@ExperimentalCoroutinesApi
+@RunWith(RobolectricTestRunner::class)
+@Config(manifest = Config.NONE)
+class MapboxTripSessionNoSetupTest {
+
+    @Test
+    fun voiceInstructionsFollowNativeStatusUpdate() {
+        // arrange
+        val voiceInstructionsObserver = VoiceInstructionsObserverRecorder()
+        val routeProgressObserver = RouteProgressObserverRecorder()
+        val nativeNavigator = FakeNativeNavigator.create()
+        val tripSession = buildTripSession(
+            nativeNavigator = nativeNavigator
+        )
+        tripSession.start(true)
+        tripSession.registerVoiceInstructionsObserver(voiceInstructionsObserver)
+        tripSession.registerRouteProgressObserver(routeProgressObserver)
+        tripSession.setRoutes(
+            listOf(createDirectionsRoute()),
+            0,
+            RoutesExtra.ROUTES_UPDATE_REASON_NEW
+        )
+
+        // act
+        nativeNavigator.updateStatus(
+            NavigationStatusOrigin.LOCATION_UPDATE,
+            createNavigationStatus(voiceInstruction = createVoiceInstruction("first"))
+        )
+        nativeNavigator.updateStatus(
+            NavigationStatusOrigin.LOCATION_UPDATE,
+            createNavigationStatus(voiceInstruction = null)
+        )
+        nativeNavigator.updateStatus(
+            NavigationStatusOrigin.LOCATION_UPDATE,
+            createNavigationStatus(voiceInstruction = createVoiceInstruction("second"))
+        )
+        // assert
+        val announcementsFromVoiceInstructionObserver = voiceInstructionsObserver.records
+            .map { it.announcement() }
+        assertEquals(listOf("first", "second"), announcementsFromVoiceInstructionObserver)
+        val announcementsFromRouteProgress = routeProgressObserver.records
+            .map { it.voiceInstructions?.announcement() }
+        assertEquals(listOf("first", "first", "second"), announcementsFromRouteProgress)
+    }
```

Comment:
The test states the same requirement as the feature description: _"RouteProgress.voiceInstruction shouldn't be null if possible. When native navigator returns null in a status update, fallback to previous value but don't trigger the new voice instruction event"_
Isn't it great. I was able to write a test that describe what I want and only then I started to think about implementation. Isn't it how TDD should work?

Reply:
Why was I able to write a test that states the same as the feature description? Classical TDD style considers tests as another client for a class. I.e. tests use a tested class just like other clients: access only public API and provide dependencies via a constructor, no interceptions somewhere in the middle. We write feature descriptions from the client's point of view as well.

That's why it's so easy to work using TDD with the classical TDD style. You write a test which is the same as the requirement. Then you think about implementation. Then you refactor without breaking and rewriting the tests.